runtime/memory.c

```diff
@@ -15,648 +15,349 @@
-  }
-}
-
-int caml_page_table_initialize(mlsize_t bytesize)
-{
-  uintnat pagesize = Page(bytesize);
-
-  caml_page_table.size = 1;
-  caml_page_table.shift = 8 * sizeof(uintnat);
-  /* Aim for initial load factor between 1/4 and 1/2 */
-  while (caml_page_table.size < 2 * pagesize) {
-    caml_page_table.size <<= 1;
-    caml_page_table.shift -= 1;
-  }
-  caml_page_table.mask = caml_page_table.size - 1;
-  caml_page_table.occupancy = 0;
-  caml_page_table.entries =
-    caml_stat_calloc_noexc(caml_page_table.size, sizeof(uintnat));
-  if (caml_page_table.entries == NULL)
-    return -1;
-  else
-    return 0;
-}
-
-static int caml_page_table_resize(void)
-{
-  struct page_table old = caml_page_table;
-  uintnat * new_entries;
-  uintnat i, h;
-
-  caml_gc_message (0x08, "Growing page table to %"
-                   ARCH_INTNAT_PRINTF_FORMAT "u entries\n",
-                   caml_page_table.size);
-
-  new_entries = caml_stat_calloc_noexc(2 * old.size, sizeof(uintnat));
-  if (new_entries == NULL) {
-    caml_gc_message (0x08, "No room for growing page table\n");
-    return -1;
-  }
-
-  caml_page_table.size = 2 * old.size;
-  caml_page_table.shift = old.shift - 1;
-  caml_page_table.mask = caml_page_table.size - 1;
-  caml_page_table.occupancy = old.occupancy;
-  caml_page_table.entries = new_entries;
-
-  for (i = 0; i < old.size; i++) {
-    uintnat e = old.entries[i];
-    if (e == 0) continue;
-    h = Hash(Page(e));
-    while (caml_page_table.entries[h] != 0)
-      h = (h + 1) & caml_page_table.mask;
-    caml_page_table.entries[h] = e;
-  }
-
-  caml_stat_free(old.entries);
-  return 0;
-}
-
-static int caml_page_table_modify(uintnat page, int toclear, int toset)
-{
-  uintnat h;
-
-  CAMLassert ((page & ~Page_mask) == 0);
-
-  /* Resize to keep load factor below 1/2 */
-  if (caml_page_table.occupancy * 2 >= caml_page_table.size) {
-    if (caml_page_table_resize() != 0) return -1;
-  }
-  h = Hash(Page(page));
-  while (1) {
-    if (caml_page_table.entries[h] == 0) {
-      caml_page_table.entries[h] = page | toset;
-      caml_page_table.occupancy++;
-      break;
-    }
-    if (Page_entry_matches(caml_page_table.entries[h], page)) {
-      caml_page_table.entries[h] =
-        (caml_page_table.entries[h] & ~toclear) | toset;
-      break;
-    }
-    h = (h + 1) & caml_page_table.mask;
-  }
-  return 0;
-}
-
-#else
-
-/* 32-bit implementation:
-   The page table is represented as a 2-level array of unsigned char */
-
-CAMLexport unsigned char * caml_page_table[Pagetable1_size];
-static unsigned char caml_page_table_empty[Pagetable2_size] = { 0, };
-
-int caml_page_table_initialize(mlsize_t bytesize)
-{
-  int i;
-  for (i = 0; i < Pagetable1_size; i++)
-    caml_page_table[i] = caml_page_table_empty;
-  return 0;
-}
-
-static int caml_page_table_modify(uintnat page, int toclear, int toset)
+/* The write barrier does not read or write the heap, it just
+   modifies domain-local data structures. */
```

Comment:
That's not true: the write barrier can call `caml_darken`.

Reply:
Agreed. Code has moved on, might once have been true. I will remove.